Support for memmap as input type

Hi, I have a problem while trying to use Regressor. Im using X_train=numpy.loadtxt() methods for loading dense data from file. This file contains a dense vector by each row. There are 525 rows with 300 dimensions each. I configured a randomizedSearch called rs. 

```
X_train: (525, 300)
y_train: (525,)
X_test: (225, 300)
Traceback (most recent call last):
  File "nn.py", line 56, in <module>
    rs.fit(X_train, y_train)
  File "/usr/local/lib/python2.7/dist-packages/sklearn/grid_search.py", line 996, in fit
    return self._fit(X, y, sampled_params)
  File "/usr/local/lib/python2.7/dist-packages/sklearn/grid_search.py", line 553, in _fit
    for parameters in parameter_iterable
  File "/usr/local/lib/python2.7/dist-packages/sklearn/externals/joblib/parallel.py", line 812, in __call__
    self.retrieve()
  File "/usr/local/lib/python2.7/dist-packages/sklearn/externals/joblib/parallel.py", line 762, in retrieve
    raise exception
sklearn.externals.joblib.my_exceptions.JoblibAttributeError: JoblibAttributeError
___________________________________________________________________________
Multiprocessing exception:
...........................................................................
/almac/ignacio/nlp-pipeline/nn.py in <module>()
     51     exit()
     52 
     53 crash = True
     54 while(crash):
     55     try:
---> 56         rs.fit(X_train, y_train)
     57         crash = False
     58     except RuntimeError:
     59         sys.stderr.write("----------- [Crashed by RunTimeERROR] --------------------- \n")
     60         crash = True

...........................................................................
/usr/local/lib/python2.7/dist-packages/sklearn/grid_search.py in fit(self=RandomizedSearchCV(cv=None, error_score='raise',...e, refit=True,
          scoring=None, verbose=0), X=array([[ -7.19239362e+05,  -2.57225021e+05,  -5....758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ]))
    991 
    992         """
    993         sampled_params = ParameterSampler(self.param_distributions,
    994                                           self.n_iter,
    995                                           random_state=self.random_state)
--> 996         return self._fit(X, y, sampled_params)
        self._fit = <bound method RandomizedSearchCV._fit of Randomi..., refit=True,
          scoring=None, verbose=0)>
        X = array([[ -7.19239362e+05,  -2.57225021e+05,  -5....758269e+00,   6.37787139e-02,  -6.47838211e-02]])
        y = array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ])
        sampled_params = <sklearn.grid_search.ParameterSampler object>
    997 
    998 
    999 
   1000 

...........................................................................
/usr/local/lib/python2.7/dist-packages/sklearn/grid_search.py in _fit(self=RandomizedSearchCV(cv=None, error_score='raise',...e, refit=True,
          scoring=None, verbose=0), X=array([[ -7.19239362e+05,  -2.57225021e+05,  -5....758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ]), parameter_iterable=<sklearn.grid_search.ParameterSampler object>)
    548         )(
    549             delayed(_fit_and_score)(clone(base_estimator), X, y, self.scorer_,
    550                                     train, test, self.verbose, parameters,
    551                                     self.fit_params, return_parameters=True,
    552                                     error_score=self.error_score)
--> 553                 for parameters in parameter_iterable
        parameters = undefined
        parameter_iterable = <sklearn.grid_search.ParameterSampler object>
    554                 for train, test in cv)
    555 
    556         # Out is a list of triplet: score, estimator, n_test_samples
    557         n_fits = len(out)

...........................................................................
/usr/local/lib/python2.7/dist-packages/sklearn/externals/joblib/parallel.py in __call__(self=Parallel(n_jobs=16), iterable=<generator object <genexpr>>)
    807             if pre_dispatch == "all" or n_jobs == 1:
    808                 # The iterable was consumed all at once by the above for loop.
    809                 # No need to wait for async callbacks to trigger to
    810                 # consumption.
    811                 self._iterating = False
--> 812             self.retrieve()
        self.retrieve = <bound method Parallel.retrieve of Parallel(n_jobs=16)>
    813             # Make sure that we get a last message telling us we are done
    814             elapsed_time = time.time() - self._start_time
    815             self._print('Done %3i out of %3i | elapsed: %s finished',
    816                         (len(self._output), len(self._output),

---------------------------------------------------------------------------
Sub-process traceback:
---------------------------------------------------------------------------
AttributeError                                     Sun Feb 21 19:22:11 2016
PID: 8977                                     Python 2.7.6: /usr/bin/python
...........................................................................
/usr/local/lib/python2.7/dist-packages/sklearn/externals/joblib/parallel.pyc in __call__(self=<sklearn.externals.joblib.parallel.BatchedCalls object>)
     67     def __init__(self, iterator_slice):
     68         self.items = list(iterator_slice)
     69         self._size = len(self.items)
     70 
     71     def __call__(self):
---> 72         return [func(*args, **kwargs) for func, args, kwargs in self.items]
     73 
     74     def __len__(self):
     75         return self._size
     76 

...........................................................................
/usr/local/lib/python2.7/dist-packages/sklearn/cross_validation.pyc in _fit_and_score(estimator=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -7.19239362e+05,  -2.57225021e+05,  -5...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ]), scorer=<function _passthrough_scorer>, train=array([175, 176, 177, 178, 179, 180, 181, 182, 1...15, 516, 517, 518, 519, 520, 521, 522, 523, 524]), test=array([  0,   1,   2,   3,   4,   5,   6,   7,  ..., 167, 168,
       169, 170, 171, 172, 173, 174]), verbose=0, parameters={'hidden0__type': 'Rectifier', 'hidden0__units': 5, 'hidden1__type': 'Rectifier', 'hidden1__units': 37, 'learning_momentum': 1.2204617397379565, 'learning_rate': 0.09480671449399372, 'learning_rule': 'rmsprop', 'output__type': 'Linear', 'regularize': None}, fit_params={}, return_train_score=False, return_parameters=True, error_score='raise')
   1526 
   1527     try:
   1528         if y_train is None:
   1529             estimator.fit(X_train, **fit_params)
   1530         else:
-> 1531             estimator.fit(X_train, y_train, **fit_params)
   1532 
   1533     except Exception as e:
   1534         if error_score == 'raise':
   1535             raise

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/mlp.pyc in fit(self=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 4.4 ,  2.2 ,  3.4 ,  4.  ,  2.6 ,  4.2 ,... 4.8 ,  4.2 ,  0.  ,  2.25,  2.6 ,  2.  ,  3.  ]), w=None)
    288         """
    289 
    290         if self.valid_set is not None:
    291             self.valid_set = self._reshape(*self.valid_set)
    292 
--> 293         return super(Regressor, self)._fit(X, y, w)
    294 
    295     def predict(self, X):
    296         """Calculate predictions for specified inputs.
    297 

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/mlp.pyc in _fit(self=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None)
    227         if self.verbose:
    228             log.debug("\nEpoch       Training Error       Validation Error       Time"
    229                       "\n------------------------------------------------------------")
    230 
    231         try:
--> 232             self._train(X, y, w)
    233         except RuntimeError as e:
    234             log.error("\n{}{}{}\n\n{}\n".format(
    235                 ansi.RED,
    236                 "A runtime exception was caught during training. This likely occurred due to\n"

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/mlp.pyc in _train(self=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None)
    146         for i in itertools.count(1):
    147             start_time = time.time()
    148             self._do_callback('on_epoch_start', locals())
    149 
    150             is_best_train = False
--> 151             avg_train_error = self._backend._train_impl(X, y, w)
    152             if avg_train_error is not None:
    153                 if math.isnan(avg_train_error):
    154                     raise RuntimeError("Training diverged and returned NaN.")
    155                 

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/backend/lasagne/mlp.pyc in _train_impl(self=<sknn.backend.lasagne.mlp.MultiLayerPerceptronBackend object>, X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None)
    300 
    301         self._print('\r')
    302         return loss / count
    303 
    304     def _train_impl(self, X, y, w=None):
--> 305         return self._batch_impl(X, y, w, self.trainer, mode='train', output='.', shuffle=True)
    306 
    307     def _valid_impl(self, X, y, w=None):
    308         return self._batch_impl(X, y, w, self.validator, mode='valid', output=' ', shuffle=False)
    309 

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/backend/lasagne/mlp.pyc in _batch_impl(self=<sknn.backend.lasagne.mlp.MultiLayerPerceptronBackend object>, X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None, processor=<theano.compile.function_module.Function object>, mode=u'train', output=u'.', shuffle=True)
    281             sys.stdout.flush()
    282 
    283     def _batch_impl(self, X, y, w, processor, mode, output, shuffle):
    284         progress, batches = 0, X.shape[0] / self.batch_size
    285         loss, count = 0.0, 0
--> 286         for Xb, yb, wb, _ in self._iterate_data(self.batch_size, X, y, w, shuffle):
    287             self._do_callback('on_batch_start', locals())
    288 
    289             if mode == 'train':
    290                 loss += processor(Xb, yb, wb if wb is not None else 1.0)

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/backend/lasagne/mlp.pyc in _iterate_data(self=<sknn.backend.lasagne.mlp.MultiLayerPerceptronBackend object>, batch_size=1, X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None, shuffle=True)
    270         if shuffle:
    271             numpy.random.shuffle(indices)
    272 
    273         for index in range(0, total_size, batch_size):
    274             excerpt = indices[index:index + batch_size]
--> 275             Xb, yb, wb = cast(X, excerpt), cast(y, excerpt), cast(w, excerpt)
    276             yield Xb, yb, wb, excerpt
    277 
    278     def _print(self, text):
    279         if self.verbose:

...........................................................................
/usr/local/lib/python2.7/dist-packages/sknn/backend/lasagne/mlp.pyc in cast(array=memmap([[  6.39665702e+06,  -1.92380979e+05,   6...693613e+00,   5.43417984e+00,  -1.42057322e+00]]), indices=array([262]))
    258             if array is None:
    259                 return None
    260 
    261             array = array[indices]
    262             if type(array) != numpy.ndarray:
--> 263                 array = array.todense()
    264             if array.dtype != theano.config.floatX:
    265                 array = array.astype(theano.config.floatX)
    266             return array
    267 
AttributeError: 'memmap' object has no attribute 'todense'
```

Please help..

Not indeed. It gives an error:
```
_64.egg/sklearn/grid_search.py", line 996, in fit
    return self._fit(X, y, sampled_params)
  File "/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/grid_search.py", line 553, in _fit
    for parameters in parameter_iterable
  File "/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/externals/joblib/parallel.py", line 812, in __call__
    self.retrieve()
  File "/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/externals/joblib/parallel.py", line 762, in retrieve
    raise exception
sklearn.externals.joblib.my_exceptions.JoblibAssertionError: JoblibAssertionError
___________________________________________________________________________
Multiprocessing exception:
...........................................................................
/home/iarroyof/nlp-pipeline/nn.py in <module>()
     58 
     59 for t in xrange(N):
     60     crash = True
     61     while(crash):
     62         try:
---> 63             rs.fit(X_train, y_train)
     64             crash = False
     65         except RuntimeError:
     66             sys.stderr.write("--------------------- [Crashed by RunTimeERROR] --------------------- \n")
     67             crash = True

...........................................................................
/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/grid_search.py in fit(self=RandomizedSearchCV(cv=None, error_score='raise',...e, refit=True,
          scoring=None, verbose=0), X=array([[ -7.19239362e+05,  -2.57225021e+05,  -5....758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ]))
    991 
    992         """
    993         sampled_params = ParameterSampler(self.param_distributions,
    994                                           self.n_iter,
    995                                           random_state=self.random_state)
--> 996         return self._fit(X, y, sampled_params)
        self._fit = <bound method RandomizedSearchCV._fit of Randomi..., refit=True,
          scoring=None, verbose=0)>
        X = array([[ -7.19239362e+05,  -2.57225021e+05,  -5....758269e+00,   6.37787139e-02,  -6.47838211e-02]])
        y = array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ])
        sampled_params = <sklearn.grid_search.ParameterSampler object>
    997 
    998 
    999 
   1000 

...........................................................................
/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/grid_search.py in _fit(self=RandomizedSearchCV(cv=None, error_score='raise',...e, refit=True,
          scoring=None, verbose=0), X=array([[ -7.19239362e+05,  -2.57225021e+05,  -5....758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ]), parameter_iterable=<sklearn.grid_search.ParameterSampler object>)
    548         )(
    549             delayed(_fit_and_score)(clone(base_estimator), X, y, self.scorer_,
    550                                     train, test, self.verbose, parameters,
    551                                     self.fit_params, return_parameters=True,
    552                                     error_score=self.error_score)
--> 553                 for parameters in parameter_iterable
        parameters = undefined
        parameter_iterable = <sklearn.grid_search.ParameterSampler object>
    554                 for train, test in cv)
    555 
    556         # Out is a list of triplet: score, estimator, n_test_samples
    557         n_fits = len(out)

...........................................................................
/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/externals/joblib/parallel.py in __call__(self=Parallel(n_jobs=16), iterable=<generator object <genexpr>>)
    807             if pre_dispatch == "all" or n_jobs == 1:
    808                 # The iterable was consumed all at once by the above for loop.
    809                 # No need to wait for async callbacks to trigger to
    810                 # consumption.
    811                 self._iterating = False
--> 812             self.retrieve()
        self.retrieve = <bound method Parallel.retrieve of Parallel(n_jobs=16)>
    813             # Make sure that we get a last message telling us we are done
    814             elapsed_time = time.time() - self._start_time
    815             self._print('Done %3i out of %3i | elapsed: %s finished',
    816                         (len(self._output), len(self._output),

---------------------------------------------------------------------------
Sub-process traceback:
---------------------------------------------------------------------------
AssertionError                                     Mon Feb 22 21:11:40 2016
PID: 18416                                    Python 2.7.6: /usr/bin/python
...........................................................................
/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/externals/joblib/parallel.pyc in __call__(self=<sklearn.externals.joblib.parallel.BatchedCalls object>)
     67     def __init__(self, iterator_slice):
     68         self.items = list(iterator_slice)
     69         self._size = len(self.items)
     70 
     71     def __call__(self):
---> 72         return [func(*args, **kwargs) for func, args, kwargs in self.items]
     73 
     74     def __len__(self):
     75         return self._size
     76 

...........................................................................
/usr/local/lib/python2.7/dist-packages/scikit_learn-0.17-py2.7-linux-x86_64.egg/sklearn/cross_validation.pyc in _fit_and_score(estimator=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -7.19239362e+05,  -2.57225021e+05,  -5...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 2.6 ,  2.6 ,  3.8 ,  4.2 ,  0.8 ,  1.4 ,... 4.2 ,  0.  ,  2.25,
        2.6 ,  2.  ,  3.  ]), scorer=<function _passthrough_scorer>, train=array([175, 176, 177, 178, 179, 180, 181, 182, 1...15, 516, 517, 518, 519, 520, 521, 522, 523, 524]), test=array([  0,   1,   2,   3,   4,   5,   6,   7,  ..., 167, 168,
       169, 170, 171, 172, 173, 174]), verbose=0, parameters={'hidden0__type': 'Rectifier', 'hidden0__units': 16, 'hidden1__type': 'Sigmoid', 'hidden1__units': 39, 'learning_momentum': 0.3153464892915808, 'learning_rate': 0.023125104835940262, 'learning_rule': 'adagrad', 'output__type': 'Softmax', 'regularize': None}, fit_params={}, return_train_score=False, return_parameters=True, error_score='raise')
   1526 
   1527     try:
   1528         if y_train is None:
   1529             estimator.fit(X_train, **fit_params)
   1530         else:
-> 1531             estimator.fit(X_train, y_train, **fit_params)
   1532 
   1533     except Exception as e:
   1534         if error_score == 'raise':
   1535             raise

...........................................................................
/home/iarroyof/scikit-neuralnetwork/sknn/mlp.pyc in fit(self=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([ 4.4 ,  2.2 ,  3.4 ,  4.  ,  2.6 ,  4.2 ,... 4.8 ,  4.2 ,  0.  ,  2.25,  2.6 ,  2.  ,  3.  ]), w=None)
    288         """
    289 
    290         if self.valid_set is not None:
    291             self.valid_set = self._reshape(*self.valid_set)
    292 
--> 293         return super(Regressor, self)._fit(X, y, w)
    294 
    295     def predict(self, X):
    296         """Calculate predictions for specified inputs.
    297 

...........................................................................
/home/iarroyof/scikit-neuralnetwork/sknn/mlp.pyc in _fit(self=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None)
    206             "Expecting same number of input and output samples."
    207         data_shape, data_size = X.shape, X.size+y.size
    208         X, y = self._reshape(X, y)
    209 
    210         if not self.is_initialized:
--> 211             X, y = self._initialize(X, y, w)
    212 
    213         log.info("Training on dataset of {:,} samples with {:,} total size.".format(data_shape[0], data_size))
    214         if data_shape[1:] != X.shape[1:]:
    215             log.warning("  - Reshaping input array from {} to {}.".format(data_shape, X.shape))

...........................................................................
/home/iarroyof/scikit-neuralnetwork/sknn/mlp.pyc in _initialize(self=Regressor(batch_size=1, callback=None, debug=Fal...e, warning=None, weight_decay=None, weights=None), X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None)
     37             "This neural network has already been initialized."
     38         self._create_specs(X, y)
     39 
     40         backend.setup()
     41         self._backend = backend.MultiLayerPerceptronBackend(self)
---> 42         return self._backend._initialize_impl(X, y, w)
     43 
     44     def _check_layer(self, layer, required, optional=[]):
     45         required.extend(['name', 'type'])
     46         for r in required:

...........................................................................
/home/iarroyof/scikit-neuralnetwork/sknn/backend/lasagne/mlp.py in _initialize_impl(self=<sknn.backend.lasagne.mlp.MultiLayerPerceptronBackend object>, X=memmap([[ -1.71586427e+07,   2.45456060e+06,   6...758269e+00,   6.37787139e-02,  -6.47838211e-02]]), y=array([[ 4.4 ],
       [ 2.2 ],
       [ 3.4 ],
...
       [ 2.6 ],
       [ 2.  ],
       [ 3.  ]]), w=None)
    232         params = []
    233         for spec, mlp_layer in zip(self.layers, self.mlp):
    234             if spec.frozen: continue
    235             params.extend(mlp_layer.get_params())
    236 
--> 237         self.trainer, self.validator = self._create_mlp_trainer(params)
        self.trainer = None
        self.validator = None
        self._create_mlp_trainer = <bound method MultiLayerPerceptronBackend._creat....lasagne.mlp.MultiLayerPerceptronBackend object>>
        params = [W, b, W, b, W, b]
    238         return X, y
    239 
    240     def _predict_impl(self, X):
    241         if self.is_convolution():

...........................................................................
/home/iarroyof/scikit-neuralnetwork/sknn/backend/lasagne/mlp.py in _create_mlp_trainer(self=<sknn.backend.lasagne.mlp.MultiLayerPerceptronBackend object>, params=[W, b, W, b, W, b])
     71         cost_symbol = self.cost_function(self.trainer_output, self.data_output)
     72         cost_symbol = lasagne.objectives.aggregate(cost_symbol.T, self.data_mask, mode='mean')
     73 
     74         if self.regularizer is not None:
     75             cost_symbol = cost_symbol + self.regularizer
---> 76         return self._create_trainer_function(params, cost_symbol)
        self._create_trainer_function = <bound method MultiLayerPerceptronBackend._creat....lasagne.mlp.MultiLayerPerceptronBackend object>>
        params = [W, b, W, b, W, b]
        cost_symbol = Elemwise{true_div,no_inplace}.0
     77 
     78     def _create_trainer_function(self, params, cost):
     79         if self.learning_rule in ('sgd', 'adagrad', 'adadelta', 'rmsprop', 'adam'):
     80             lr = getattr(lasagne.updates, self.learning_rule)

...........................................................................
/home/iarroyof/scikit-neuralnetwork/sknn/backend/lasagne/mlp.py in _create_trainer_function(self=<sknn.backend.lasagne.mlp.MultiLayerPerceptronBackend object>, params=[W, b, W, b, W, b], cost=Elemwise{true_div,no_inplace}.0)
     88                 "Learning rule type `%s` is not supported." % self.learning_rule)
     89 
     90         trainer = theano.function([self.data_input, self.data_output, self.data_mask], cost,
     91                                    updates=self._learning_rule,
     92                                    on_unused_input='ignore',
---> 93                                    allow_input_downcast=True)
     94 
     95         compare = self.cost_function(self.network_output, self.data_correct).mean()
     96         validator = theano.function([self.data_input, self.data_correct], compare,
     97                                     allow_input_downcast=True)

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/compile/function.pyc in function(inputs=[X, y, m], outputs=Elemwise{true_div,no_inplace}.0, mode=None, updates=OrderedDict([(<TensorType(float64, matrix)>, Ele...no_inplace}.0), (b, Elemwise{sub,no_inplace}.0)]), givens=[], no_default_updates=False, accept_inplace=False, name='/home/iarroyof/scikit-neuralnetwork/sknn/backend/lasagne/mlp.py:93', rebuild_strict=True, allow_input_downcast=True, profile=None, on_unused_input=u'ignore')
    261                 no_default_updates=no_default_updates,
    262                 accept_inplace=accept_inplace, name=name,
    263                 rebuild_strict=rebuild_strict,
    264                 allow_input_downcast=allow_input_downcast,
    265                 on_unused_input=on_unused_input,
--> 266                 profile=profile)
    267     # We need to add the flag check_aliased inputs if we have any mutable or
    268     # borrowed used defined inputs
    269     fn._check_for_aliased_inputs = check_for_aliased_inputs
    270     return fn

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/compile/pfunc.pyc in pfunc(params=[X, y, m], outputs=Elemwise{true_div,no_inplace}.0, mode=None, updates=OrderedDict([(<TensorType(float64, matrix)>, Ele...no_inplace}.0), (b, Elemwise{sub,no_inplace}.0)]), givens=[], no_default_updates=False, accept_inplace=False, name='/home/iarroyof/scikit-neuralnetwork/sknn/backend/lasagne/mlp.py:93', rebuild_strict=True, allow_input_downcast=True, profile=False, on_unused_input=u'ignore')
    506                     mutable=False, borrow=True, shared=True)
    507         inputs.append(si)
    508 
    509     return orig_function(inputs, cloned_outputs, mode,
    510             accept_inplace=accept_inplace, name=name, profile=profile,
--> 511             on_unused_input=on_unused_input)
    512 
    513 
    514 def _pfunc_param_to_in(param, strict=False, allow_downcast=None):
    515     if isinstance(param, Constant):

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/compile/function_module.pyc in orig_function(inputs=[In(X), In(y), In(m), In(W -> Elemwise{sub,no_inplace}.0), In(b -> Elemwise{sub,no_inplace}.0), In(W -> Elemwise{sub,no_inplace}.0), In(b -> Elemwise{sub,no_inplace}.0), In(W -> Elemwise{sub,no_inplace}.0), In(b -> Elemwise{sub,no_inplace}.0), In(<TensorType(float64, matrix)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, vector)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, matrix)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, vector)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, matrix)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, vector)> -> Elemwise{add,no_inplace}.0)], outputs=Out(Elemwise{true_div,no_inplace}.0,False), mode=<theano.compile.mode.Mode object>, accept_inplace=False, name='/home/iarroyof/scikit-neuralnetwork/sknn/backend/lasagne/mlp.py:93', profile=False, on_unused_input=u'ignore')
   1460         fn = Maker(inputs,
   1461                    outputs,
   1462                    mode,
   1463                    accept_inplace=accept_inplace,
   1464                    profile=profile,
-> 1465                    on_unused_input=on_unused_input).create(
   1466                        defaults)
   1467 
   1468     t2 = time.time()
   1469     if profile:

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/compile/function_module.pyc in __init__(self=<theano.compile.function_module.FunctionMaker object>, inputs=[In(X), In(y), In(m), In(W -> Elemwise{sub,no_inplace}.0), In(b -> Elemwise{sub,no_inplace}.0), In(W -> Elemwise{sub,no_inplace}.0), In(b -> Elemwise{sub,no_inplace}.0), In(W -> Elemwise{sub,no_inplace}.0), In(b -> Elemwise{sub,no_inplace}.0), In(<TensorType(float64, matrix)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, vector)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, matrix)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, vector)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, matrix)> -> Elemwise{add,no_inplace}.0), In(<TensorType(float64, vector)> -> Elemwise{add,no_inplace}.0)], outputs=[Out(Elemwise{true_div,no_inplace}.0,False)], mode=<theano.compile.mode.Mode object>, accept_inplace=False, function_builder=<class 'theano.compile.function_module.Function'>, profile=None, on_unused_input=u'ignore', fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)])
   1155                 # now optimize the graph
   1156                 if theano.config.cache_optimizations:
   1157                     optimizer_profile = self.optimize_graph_with_cache(
   1158                         optimizer, inputs, outputs)
   1159                 else:    
-> 1160                     optimizer_profile = optimizer(fgraph)
   1161                     
   1162                 end_optimizer = time.time()
   1163                 opt_time = end_optimizer - start_optimizer
   1164                 if profile:

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in __call__(self=[<theano.tensor.opt.ShapeOptimizer object at 0x7...gof.opt.MergeOptimizer object at 0x7f39d2791d10>], fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)])
     85 
     86     def __call__(self, fgraph):
     87         """WRITEME
     88         Same as self.optimize(fgraph)
     89         """
---> 90         return self.optimize(fgraph)
     91 
     92     def add_requirements(self, fgraph):
     93         """WRITEME
     94         Add features to the fgraph that are required to apply the optimization.

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in optimize(self=[<theano.tensor.opt.ShapeOptimizer object at 0x7...gof.opt.MergeOptimizer object at 0x7f39d2791d10>], fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)], *args=(), **kwargs={})
     76         """
     77         self.add_requirements(fgraph)
     78         try:
     79             orig = theano.tensor.basic.constant.enable
     80             theano.tensor.basic.constant.enable = False
---> 81             ret = self.apply(fgraph, *args, **kwargs)
     82         finally:
     83             theano.tensor.basic.constant.enable = orig
     84         return ret
     85 

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in apply(self=[<theano.tensor.opt.ShapeOptimizer object at 0x7...gof.opt.MergeOptimizer object at 0x7f39d2791d10>], fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)])
    190         nb_node_before = len(fgraph.apply_nodes)
    191         sub_profs = []
    192         for optimizer in self:
    193             try:
    194                 t0 = time.time()
--> 195                 sub_prof = optimizer.optimize(fgraph)
    196                 l.append(float(time.time() - t0))
    197                 sub_profs.append(sub_prof)
    198                 if fgraph.profile:
    199                     sub_validate_time.append(fgraph.profile.validate_time)

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in optimize(self=<theano.gof.opt.EquilibriumOptimizer object>, fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)], *args=(), **kwargs={})
     76         """
     77         self.add_requirements(fgraph)
     78         try:
     79             orig = theano.tensor.basic.constant.enable
     80             theano.tensor.basic.constant.enable = False
---> 81             ret = self.apply(fgraph, *args, **kwargs)
     82         finally:
     83             theano.tensor.basic.constant.enable = orig
     84         return ret
     85 

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in apply(self=<theano.gof.opt.EquilibriumOptimizer object>, fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)], start_from=[Elemwise{true_div,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0, Elemwise{add,no_inplace}.0, Elemwise{add,no_inplace}.0, Elemwise{add,no_inplace}.0, Elemwise{add,no_inplace}.0, Elemwise{add,no_inplace}.0, Elemwise{add,no_inplace}.0])
   1832                     for lopt in (self.local_optimizers_all +
   1833                                  self.local_optimizers_map.get(type(node.op), []) +
   1834                                  self.local_optimizers_map.get(node.op, [])):
   1835                         nb = change_tracker.nb_imported
   1836                         t_opt = time.time()
-> 1837                         lopt_change = self.process_node(fgraph, node, lopt)
   1838                         time_opts[lopt] += time.time() - t_opt
   1839                         if lopt_change:
   1840                             process_count.setdefault(lopt, 0)
   1841                             process_count[lopt] += 1

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in process_node(self=<theano.gof.opt.EquilibriumOptimizer object>, fgraph=[*3 -> Elemwise{true_div,no_inplace}(Elemwise{tr...,no_inplace}(<TensorType(float64, vector)>, *16)], node=DimShuffle{x}(TensorConstant{2.0}), lopt=<theano.gof.opt.FromFunctionLocalOptimizer object>)
   1493             replacements = lopt.transform(node)
   1494         except Exception, e:
   1495             if self.failure_callback is not None:
   1496                 self.failure_callback(e, self,
   1497                                       [(x, None) for x in node.outputs],
-> 1498                                       lopt)
   1499                 return False
   1500             else:
   1501                 raise
   1502         if replacements is False or replacements is None:

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in warn_inplace(exc=AssertionError(), nav=<theano.gof.opt.EquilibriumOptimizer object>, repl_pairs=[(DimShuffle{x}.0, None)], local_opt=<theano.gof.opt.FromFunctionLocalOptimizer object>)
   1391 
   1392         ignore InconsistencyErrors, print traceback
   1393         """
   1394         if isinstance(exc, InconsistencyError):
   1395             return
-> 1396         return NavigatorOptimizer.warn(exc, nav, repl_pairs, local_opt)
   1397 
   1398     @staticmethod
   1399     def warn_ignore(exc, nav, repl_pairs, local_opt):
   1400         """failure_callback for NavigatorOptimizer: ignore all errors

...........................................................................
/usr/local/lib/python2.7/dist-packages/theano/gof/opt.pyc in warn(exc=AssertionError(), nav=<theano.gof.opt.EquilibriumOptimizer object>, repl_pairs=[(DimShuffle{x}.0, None)], local_opt=<theano.gof.opt.FromFunctionLocalOptimizer object>)
   1381         if config.on_opt_error == 'pdb':
   1382             pdb.post_mortem(sys.exc_info()[2])
   1383         elif isinstance(exc, AssertionError) or config.on_opt_error == 'raise':
   1384             # We always crash on AssertionError because something may be
   1385             # seriously wrong if such an exception is raised.
-> 1386             raise exc
   1387 
   1388     @staticmethod
   1389     def warn_inplace(exc, nav, repl_pairs, local_opt):
   1390         """failure_callback for NavigatorOptimizer

AssertionError: 
```
I suppose it is absolutely necessary carting any resulting data type to numpy array.

--- a/sknn/backend/lasagne/mlp.py
+++ b/sknn/backend/lasagne/mlp.py
@@ -259,7 +259,8 @@ class MultiLayerPerceptronBackend(BaseBackend):
                 return None
 
             array = array[indices]
-            if type(array) != numpy.ndarray:
+            if not isinstance(array, numpy.ndarray):
+                assert hasattr(array, 'todense'), "Unknown data format and cannot convert to numpy.ndarray."
                 array = array.todense()
             if array.dtype != theano.config.floatX:
                 array = array.astype(theano.config.floatX)

--- a/sknn/tests/test_data.py
+++ b/sknn/tests/test_data.py
@@ -101,7 +101,7 @@ class TestMaskedDataRegression(unittest.TestCase):
         # Make sure the examples weighted 1.0 have low error, 0.0 high error.
         masked = abs(a_out - v_out).T * a_mask
         print('masked', masked)
-        assert_true((masked < 4.0E-1).all())
+        assert_true((masked < 5.0E-1).all())
         inversed = abs(a_out - v_out).T * (1.0 - a_mask)
         print('inversed', inversed)
         assert_greater(inversed.mean(), masked.mean())

--- a/sknn/tests/test_types.py
+++ b/sknn/tests/test_types.py
@@ -1,7 +1,10 @@
 import unittest
 from nose.tools import (assert_is_not_none, assert_raises, assert_equal, assert_true)
 
+import os
 import random
+import shutil
+import tempfile
 
 import theano
 import numpy
@@ -72,6 +75,33 @@ class TestScipySparseMatrix(unittest.TestCase):
             X = sparse_matrix((8, 4), dtype=numpy.float32)
             yp = self.nn._predict(X)
             assert_equal(yp.dtype, numpy.float32)
+
+
+class TestMemoryMap(unittest.TestCase):
+
+    def setUp(self):
+        self.nn = MLP(layers=[L("Linear", units=3)], n_iter=1)
+        self.directory = tempfile.mkdtemp()
+
+    def tearDown(self):
+        shutil.rmtree(self.directory)
+
+    def make(self, name, shape, dtype):
+        filename = os.path.join(self.directory, name)
+        return numpy.memmap(filename, dtype=dtype, mode='w+', shape=shape)
+
+    def test_FitAllTypes(self):
+        for t in ['float32', 'float64']:
+            theano.config.floatX = t
+            X = self.make('X', (12, 3), dtype=t)
+            y = self.make('y', (12, 3), dtype=t)
+            self.nn._fit(X, y)
+
+    def test_PredictAllTypes(self):
+        for t in ['float32', 'float64']:
+            theano.config.floatX = t
+            X = self.make('X', (12, 3), dtype=t)
+            yp = self.nn._predict(X)
 
 
 class TestConvolution(unittest.TestCase):

